### PR TITLE
Add support for reading Fortinet firmware files

### LINKED ISSUE #655: Add support for reading Fortinet firmware files
Would be nice to have the ability to mount or dump Fortinet/FortiGate firmware files.

These firmware files are generally:

* gzip compressed
* obfuscated using a custom XOR algorithm

Would be nice to directly able to use `target-shell` on them without any extra steps.

A public mirror of some FortiGate firmware images can be found here: https://dl.partian.co/FortiGate/

### PR DESCRIPTION
**Pull request by yunzheng**
This adds transparant decompression and deobfuscation of Fortinet firmware files via the `FortiFirmwareContainer` and `FortiFirmwareFile` classes.

The filename has to end with `-FORTINET.out` for it to be detected as a Fortinet firmware file.

Example usage:

    $ target-shell FGT_VM64-v7.4.3.F-build2573-FORTINET.out
    $ target-fs FGT_2000E-v5-build1673-FORTINET.out ls /migadmin

Or to decompress and deobfuscate a firmware file to stdout:

    $ python3 -m dissect.target.containers.fortifw <FIRMWARE FILE> | xxd

Closes #655

### CODE REVIEW

File: dissect/target/containers/fortifw.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+logger = logging.getLogger(__name__)
+
+
+def find_xor_key(fobj: io.BytesIO) -> bytes:
+    """Find the XOR key for the firmware file by using known plaintext of zeros.
+
+    File object ``fobj`` should be at the correct offset where it should decode to all zeroes (0x00).
+
+    Arguments:
+        fobj: file object to read from
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:24]:
```suggestion
        fh: File-like object to read from.
```

File: dissect/target/containers/fortifw.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+    Returns:
+        bytes: XOR key, zero bytes if no key is found
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:27]:
```suggestion
        bytes: XOR key, zero bytes if no key is found.
```

File: dissect/target/containers/fortifw.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+from dissect.util.stream import RangeStream, RelativeStream
+
+from dissect.target.container import Container
+from dissect.target.tools.utils import catch_sigpipe
+
+logger = logging.getLogger(__name__)
+
+
+def find_xor_key(fobj: io.BytesIO) -> bytes:
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:18]:
```suggestion
def find_xor_key(fh: io.BytesIO) -> bytes:
```
`fh` is a bit more consistent with other code.

File: dissect/target/containers/fortifw.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    pos = fobj.tell()
+    buf = fobj.read(32)
+    fobj.seek(pos)
+
+    if pos % 512 == 0:
+        xor_char = 0xFF
+    else:
+        fobj.seek(pos - 1)
+        xor_char = ord(fobj.read(1))
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:39]:
```suggestion
    pos = fh.tell()
    buf = fh.read(32)
    fh.seek(pos)

    if pos % 512 == 0:
        xor_char = 0xFF
    else:
        fobj.seek(pos - 1)
        xor_char = ord(fh.read(1))
```

File: dissect/target/containers/fortifw.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    koffset = 32 - (pos & 0x1F)
+    key = islice(cycle(key), koffset, koffset + 32)
+    return bytes(key)
+
+
+class FortiFirmwareFile:
+    """Fortinet firmware file, handles transparant decompression and deobfuscation of the firmware file."""
+
+    def __init__(self, fobj: io.BytesIO) -> None:
+        self.fh = fobj
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:56]:
```suggestion
    def __init__(self, fh: BinaryIO):
        self.fh = fh
```

File: dissect/target/containers/fortifw.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+            while True:
+                data = self.fh.read(io.DEFAULT_BUFFER_SIZE)
+                if not data:
+                    break
+                d = dec.decompress(dec.unconsumed_tail + data)
+                self.size += len(d)
+
+            # Ignore the trailer data of the gzip file if we have any
+            if dec.unused_data:
+                self.fh.seek(-len(dec.unused_data), os.SEEK_END)
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:80]:
```suggestion
                self.fh.seek(-len(dec.unused_data), io.SEEK_END)
```

File: dissect/target/containers/fortifw.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+                self.trailer_offset = self.fh.tell()
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:81]: Can you initialize these class attributes to `None` if the file is not compressed?

File: dissect/target/containers/fortifw.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+from itertools import cycle, islice
+from pathlib import Path
+from typing import BinaryIO, Optional, Sequence, Union
+
+from dissect.util.stream import RangeStream, RelativeStream
+
+from dissect.target.container import Container
+from dissect.target.tools.utils import catch_sigpipe
+
+logger = logging.getLogger(__name__)
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:15]:
```suggestion
log = logging.getLogger(__name__)
```

File: dissect/target/containers/fortifw.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+        logger.info("firmware size: %s", self.size)
+        logger.info("key: %r", self.xor_key)
+        logger.info("gzipped: %s", self.is_gzipped)
+        self.fh.seek(0)
+
+    def seek(self, offset, whence=io.SEEK_SET):
+        return self.fh.seek(offset, whence)
+
+    def read(self, n=-1):
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:114]: Have you tried turning this class into a subclass of `AlignedStream`?

File: dissect/target/containers/fortifw.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
+import gzip
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:1]:
```suggestion
from __future__ import annotations

import gzip
```

File: dissect/target/containers/fortifw.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+        if n == -1:
+            n = None
+        return bytes(data[:n])
+
+
+class FortiFirmwareContainer(Container):
+    __type__ = "fortifw"
+
+    def __init__(self, fh: Union[BinaryIO, Path], *args, **kwargs) -> None:
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:151]:
```suggestion
    def __init__(self, fh: BinaryIO | Path, *args, **kwargs):
```

File: dissect/target/containers/fortifw.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+        # Open the firmware file
+        self.ff = FortiFirmwareFile(fh)
+
+        # seek to MBR
+        self.fw = RelativeStream(self.ff, 0x200)
+        super().__init__(self.fw, self.ff.size, *args, **kwargs)
+
+    @staticmethod
+    def detect_fh(fh: BinaryIO, original: Union[list, BinaryIO]) -> bool:
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:163]:
```suggestion
    def detect_fh(fh: BinaryIO, original: list | BinaryIO) -> bool:
```

File: dissect/target/containers/fortifw.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+        return False
+
+    @staticmethod
+    def detect_path(path: Path, original: Union[list, BinaryIO]) -> bool:
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:167]:
```suggestion
    def detect_path(path: Path, original: list | BinaryIO) -> bool:
```

File: dissect/target/containers/fortifw.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+
+
+@catch_sigpipe
+def main(argv: Optional[Sequence[str]] = None) -> None:
+    import argparse
+    import sys
+
+    parser = argparse.ArgumentParser(description="Decompress and deobfuscate Fortinet firmware file to stdout.")
+    parser.add_argument("file", type=argparse.FileType("rb"), help="Fortinet firmware file")
+    parser.add_argument("--verbose", "-v", action="store_true", help="Verbose output")
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:191]:
```suggestion
    parser.add_argument("--verbose", "-v", action="store_true", help="verbose output")
```

File: dissect/target/containers/fortifw.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+
+    if args.verbose:
+        logging.basicConfig(level=logging.INFO)
+
+    ff = FortiFirmwareFile(args.file)
+    while True:
+        data = ff.read(io.DEFAULT_BUFFER_SIZE)
+        if not data:
+            break
+        sys.stdout.buffer.write(data)
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:202]: You can maybe use `shutil.copyfileobj(ff, sys.stdout.buffer)`

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/containers/fortifw.py (hunk 9 above)
Comment (yunzheng) [reply dissect/target/containers/fortifw.py:114]: what benefit does that give?

File: dissect/target/containers/fortifw.py (hunk 15 above)
Comment (yunzheng) [reply dissect/target/containers/fortifw.py]: nice one, that saves some lines :)

File: dissect/target/containers/fortifw.py (hunk 7 above)
Comment (yunzheng) [reply dissect/target/containers/fortifw.py:84]: done

File: dissect/target/containers/fortifw.py (hunk 9 above)
Comment (Schamper) [reply dissect/target/containers/fortifw.py:117]: Some caching, read offset is always aligned, real compatibility with `io.RawIOBase`.

Comment (yunzheng) [reply dissect/target/containers/fortifw.py]: Fixed, AlignedStream is awesome

File: dissect/target/containers/fortifw.py [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+                log.info("Found key %r @ offset %s", self.xor_key, zero_offset)
+                break
+        else:
+            self.xor_key = None
+            log.info("No xor key found")
+
+        # Determine the size of the firmware file if we didn't calculate it yet
+        if size is None:
+            self.fh.seek(0, io.SEEK_END)
+            size = self.fh.tell()
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:108]:
```suggestion
            size = self.fh.seek(0, io.SEEK_END)
```

File: dissect/target/containers/fortifw.py [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+
+    def tell(self) -> int:
+        return self.fw.tell()
+
+    def close(self) -> None:
+        pass
+
+
+@catch_sigpipe
+def main(argv: Optional[Sequence[str]] = None) -> None:
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:174]:
```suggestion
def main(argv: Optional[list[str]] = None) -> None:
```
I know `Sequence` is technically more correct, but `list` is more consistent across the code base atm.

File: dissect/target/containers/fortifw.py [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+
+from dissect.util.stream import AlignedStream, RangeStream, RelativeStream
+
+from dissect.target.container import Container
+from dissect.target.tools.utils import catch_sigpipe
+
+log = logging.getLogger(__name__)
+
+
+def find_xor_key(fh: io.BytesIO) -> bytes:
```
Comment (Schamper) [inline dissect/target/containers/fortifw.py:19]:
```suggestion
def find_xor_key(fh: BinaryIO) -> bytes:
```

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/containers/fortifw.py (hunk 16 above)
Comment (yunzheng) [reply dissect/target/containers/fortifw.py]: neat

**Review by Schamper — APPROVED** [review]

**Pull request merged**